src/controllers/events_controller.ts

```diff
@@ -1,42 +1,147 @@
+      openEventStream(request, response);
     } else {
-      response.status(200).json(Events.getGatewayEvents());
+      sendEventLog(request, response);
     }
   });
 
   /**
-   * Handle getting a list of events.
+   * Handle getting events of a specific type.
    */
   controller.get('/:eventName', (request, response) => {
+    // Serve either an event stream or a log depending on requested content type
+    if (request.accepts('text/event-stream')) {
+      openEventStream(request, response);
+    } else {
+      sendEventLog(request, response);
+    }
+  });
+
+  /**
+   * Open a Server-Sent Events event stream to push events to the client.
+   *
+   * @param {express.Request} request
+   * @param {express.Response} response
+   * @return {Promise}
+   */
+  async function openEventStream(
+    request: express.Request,
+    response: express.Response
+  ): Promise<void> {
+    const thingID = request.params.thingId;
     const eventName = request.params.eventName;
+    let thing: Thing | undefined;
+
+    // Don't allow event streams for events not associated with a Thing
+    if (!thingID) {
+      response.status(406).send();
+      return;
+    }
+
+    // Check requested thing exists
+    try {
+      thing = await Things.getThing(thingID);
+    } catch (error: unknown) {
+      console.error(`Thing not found ${error}`);
+      response.status(404).send();
+      return;
+    }
+
+    // Check that requested event type (if any) exists
+    if (eventName && !thing.getEvents()[eventName]) {
+      response.status(404).send();
+      return;
+    }
 
+    // Keep the socket open
+    request.socket.setKeepAlive(true);
+    // Prevent Nagle's algorithm from trying to optimise throughput
+    request.socket.setNoDelay(true);
+    // Disable inactivity timeout on the socket
+    request.socket.setTimeout(0);
+
+    // Set event stream content type
+    response.setHeader('Content-Type', 'text/event-stream');
+    // Disable caching and compression
+    response.setHeader('Cache-Control', 'no-cache,no-transform');
+    // Tell client to keep the connection alive
+    response.setHeader('Connection', 'keep-alive');
+    // Set 200 OK response
+    response.status(200);
+    // Send headers to complete the connection, but don't end the response
+    response.flushHeaders();
+
+    /**
+     * Handle an event emitted by a Thing
+     *
+     * @param {Event} event
+     */
+    function onEvent(event: Event): void {
+      // If subscribed to a particular event, filter others out
+      if (eventName && eventName != event.getName()) {
+        return;
+      }
+
+      // Generate an ID for the event
+      const eventId = Date.now();
+
+      // Push event to client via event stream
+      response.write(`id: ${eventId}\n`);
+      response.write(`event: ${event.getName()}\n`);
+      response.write(`data: ${JSON.stringify(event.getData())}\n\n`);
+    }
+
+    // Subscribe to events from the specified Thing
+    if (thing) {
```

Comment:
Ah yes, you're correct.

It gets a bit confusing because the gateway can have events which are associated with itself rather than a thing. I decided not to provide an SSE endpoint for those since we'll probably want to handle that if and when we implement the Directory Service API, but I forgot to remove the conditional after adding those checks.